flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/outlier/StochasticOutlierSelection.scala

```diff
@@ -0,0 +1,367 @@
+    */
+  def setMaxIterations(maxIterations: Int): StochasticOutlierSelection = {
+    require(maxIterations > 0, "Maximum iterations must be positive.")
+    parameters.add(MaxIterations, maxIterations)
+    this
+  }
+
+}
+
+object StochasticOutlierSelection extends WithParameters {
+
+  // ========================================= Parameters ==========================================
+  case object Perplexity extends Parameter[Double] {
+    val defaultValue: Option[Double] = Some(30)
+  }
+
+  case object ErrorTolerance extends Parameter[Double] {
+    val defaultValue: Option[Double] = Some(1e-20)
+  }
+
+  case object MaxIterations extends Parameter[Int] {
+    val defaultValue: Option[Int] = Some(5000)
+  }
+
+  // ==================================== Factory methods ==========================================
+
+  def apply(): StochasticOutlierSelection = {
+    new StochasticOutlierSelection()
+  }
+
+  // ===================================== Operations ==============================================
+  case class BreezeLabeledVector(idx: Int, data: BreezeVector[Double])
+
+  implicit val transformLabeledVectors = {
+    new TransformDataSetOperation[StochasticOutlierSelection, LabeledVector, (Int, Double)] {
+      override def transformDataSet(instance: StochasticOutlierSelection,
+                                    transformParameters: ParameterMap,
+                                    input: DataSet[LabeledVector]): DataSet[(Int, Double)] = {
+
+        val resultingParameters = instance.parameters ++ transformParameters
+
+        val vectorsWithIndex = input.map(labeledVector => {
+          BreezeLabeledVector(labeledVector.label.toInt, labeledVector.vector.asBreeze)
+        })
+
+        // Don't map back to a labeled-vector since the output of the algorithm is
+        // a single double instead of vector
+        outlierSelection(vectorsWithIndex, resultingParameters)
+      }
+    }
+  }
+
+  /** [[TransformDataSetOperation]] applies the stochastic outlier selection algorithm on a
+    * [[Vector]] which will transform the high-dimensionaly input to a single Double output.
+    *
+    * @tparam T Type of the input and output data which has to be a subtype of [[Vector]]
+    * @return [[TransformDataSetOperation]] a single double which represents the oulierness of
+    *         the input vectors, where the output is in [0, 1]
+    */
+  implicit def transformVectors[T <: Vector : BreezeVectorConverter : TypeInformation : ClassTag]
+  = {
+    new TransformDataSetOperation[StochasticOutlierSelection, T, Double] {
+      override def transformDataSet(instance: StochasticOutlierSelection,
+                                    transformParameters: ParameterMap,
+                                    input: DataSet[T]): DataSet[Double] = {
+
+        val resultingParameters = instance.parameters ++ transformParameters
+
+        // Map to the right format
+        val vectorsWithIndex = input.zipWithIndex.map(vector => {
```

Comment:
Does it strictly have to be the index you're zipping with? Would be a unique ID be enough? If so, then we could use the `DatatSetUtils.zipWithUniqueId` method which is considerably cheaper than `zipWithIndex`.